{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Date: 20.3.2023\n",
    "# Author: Antti Kiviaho\n",
    "#\n",
    "# Notebook for analysing and visualizing visium data after copy number variation, single cell mapping and clustering\n",
    "# analyses. This is the main results notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "os.chdir('/lustre/scratch/kiviaho/prostate_spatial/')\n",
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import squidpy as sq\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "\n",
    "#from cell2location.utils import select_slide\n",
    "#from cell2location.plt import plot_spatial\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from scripts.utils import load_from_pickle, get_sample_ids\n",
    "import matplotlib as mpl\n",
    "\n",
    "sc.set_figure_params(figsize=(4.7,4.7),dpi_save=1200,frameon=False,format='pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download data and format cell2location mapping results into obs columns in both aggregated adata and individual slides\n",
    "\n",
    "adata_vis = sc.read_h5ad('./c2l-results/cell2location_map_20230322/visium_adata_with_c2l_mapping_20230322.h5ad')\n",
    "\n",
    "adata_vis.obs.joint_leiden_clusters = adata_vis.obs.sample_id.astype(str) + '_' + adata_vis.obs.joint_leiden_clusters.astype(str)\n",
    "adata_vis.obs['joint_leiden_clusters'] = pd.Categorical(adata_vis.obs['joint_leiden_clusters'])\n",
    "\n",
    "\n",
    "adata_slides = load_from_pickle('./data/individual_sections_normalized_clustered.pickle')\n",
    "samples = get_sample_ids()\n",
    "\n",
    "# Copy obsm (cell2location results) to adata object obs\n",
    "adata_vis.obs[adata_vis.uns['mod']['factor_names']] = adata_vis.obsm['q05_cell_abundance_w_sf']\n",
    "\n",
    "# Copy obsm (cell2location results) to individual slides\n",
    "for sample in samples:\n",
    "    vis_subset = adata_vis[adata_vis.obs['sample_id']==sample]\n",
    "    \n",
    "    if (vis_subset.obs_names == adata_slides[sample].obs_names).all():\n",
    "        adata_slides[sample].obsm = vis_subset.obsm.copy()\n",
    "        \n",
    "        # add 5% quantile, representing confident cell abundance, 'at least this amount is present',\n",
    "        # to adata.obs with nice names for plotting\n",
    "        adata_slides[sample].uns['mod'] = vis_subset.uns['mod'].copy()\n",
    "        adata_slides[sample].obs[adata_slides[sample].uns['mod']['factor_names']] = adata_slides[sample].obsm['q05_cell_abundance_w_sf']\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SINGLE-CELL REFERENCE PLOTTING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the single-cell cell type reference: export estimated expression in 'cell type'\n",
    "adata_ref = sc.read_h5ad('/lustre/scratch/kiviaho/prostate_spatial/c2l-results/cell2location_map_20230322/sc_reference_signatures_20230322.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref.obs['dataset'].value_counts(\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_variable = 'detailed_celltypes' # phenotype\n",
    "\n",
    "meta = adata_ref.obs[[y_variable,'dataset']]\n",
    "meta['count'] = 1\n",
    "\n",
    "grouped_meta = meta.groupby(['dataset',y_variable],axis=0).sum()\n",
    "grouped_meta.reset_index(inplace=True)\n",
    "grouped_meta = grouped_meta.pivot(index='dataset',columns=y_variable,values='count')\n",
    "grouped_meta = grouped_meta.loc[['hirz_2023','chen_2021','cheng_2022','dong_2020','chen_2022','song_2022']]\n",
    "#grouped_meta.set_index('dataset',inplace=True)\n",
    "#grouped_meta\n",
    "fig,ax = plt.subplots()\n",
    "\n",
    "grouped_meta.plot.bar(stacked=True,grid=False,ax=ax,figsize=(3,5),legend=False,width=0.8 #yticks=(0,2e4,4e4,14e4),\n",
    "                      )\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/single-cell-barplot-celltypes.pdf',dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import cycle, islice\n",
    "\n",
    "y_variable = 'phenotype' # phenotype\n",
    "\n",
    "\n",
    "meta = adata_ref.obs[[y_variable,'dataset']]\n",
    "meta['count'] = 1\n",
    "\n",
    "grouped_meta = meta.groupby(['dataset',y_variable],axis=0).sum()\n",
    "grouped_meta.reset_index(inplace=True)\n",
    "grouped_meta = grouped_meta.pivot(index='dataset',columns=y_variable,values='count')\n",
    "grouped_meta = grouped_meta.loc[['hirz_2023','chen_2021','cheng_2022','dong_2020','chen_2022','song_2022']]\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "\n",
    "\n",
    "# Make a list by cycling through the colors you care about\n",
    "# to match the length of your data.\n",
    "my_colors = list(islice(cycle(['r', 'b', 'g',]), None, len(grouped_meta)))\n",
    "\n",
    "\n",
    "grouped_meta.plot.bar(stacked=True,grid=False,ax=ax,figsize=(3,5),width=0.8,color=my_colors #yticks=(0,2e4,4e4,14e4),legend=False\n",
    "                      )\n",
    "#plt.axis('off')\n",
    "#plt.savefig('./figures/single-cell-barplot-phenotypes.pdf',dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pl.umap(adata_ref,color='detailed_celltypes',size=10,legend_loc=None,\n",
    "           save='single-cell-dataset-celltypes.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(8,8))\n",
    "sc.pl.umap(adata_ref,color='dataset',size=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(8,8))\n",
    "sc.pl.umap(adata_ref,color='sample',size=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ribosomal_markers = ['RPSA','RPS2','RPS3','RPS3A','RPS4X','RPS4Y1','RPS4Y2','RPS5','RPS6','RPS7','RPS8','RPS9','RPS10','RPS11',\n",
    "'RPS12','RPS13','RPS14','RPS15','RPS15A','RPS16','RPS17','RPS18','RPS19','RPS20','RPS21','RPS23','RPS24','RPS25'\n",
    ",'RPS26','RPS27','RPS27A','RPS27L','RPS28','RPS29','FAU','RPLP0','RPLP1','RPLP2','RPL3','RPL3L','RPL4','RPL5','RPL6',\n",
    "'RPL7','RPL7A','RPL7L1','RPL8','RPL9','RPL10','RPL10A','RPL10L','RPL11','RPL12','RPL13','RPL13A','RPL14','RPL15','RPL17',\n",
    "'RPL18','RPL18A','RPL19','RPL21','RPL22','RPL22L1','RPL23','RPL23A','RPL24','RPL26','RPL26L1','RPL27','RPL27A','RPL28','RPL29',\n",
    "'RPL30','RPL31','RPL32','RPL34','RPL35','RPL35A','RPL36','RPL36A','RPL36AL','RPL37','RPL37A','RPL38','RPL39','RPL39L','UBA52','RPL41']\n",
    "sc.tl.score_genes(adata_ref,ribosomal_markers,score_name='ribosomal_score')\n",
    "sc.pl.umap(adata_ref,color=['ribosomal_score'],size=10)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visium mapping plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(4.7,4.7),dpi_save=1200,frameon=False,format='pdf')\n",
    "\n",
    "sc.tl.umap(adata_slides[sample])\n",
    "sc.pl.umap(adata_slides[sample],color='joint_leiden_clusters',size=30,legend_loc=None,\n",
    "           save='_sample_'+sample+'.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(4.7,4.7),dpi_save=1200,frameon=False,format='pdf')\n",
    "\n",
    "sc.pl.spatial(adata_slides[sample],color='joint_leiden_clusters',size=1.3,legend_loc=None,\n",
    "           save='_sample_'+sample+'.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plots cell type abundance violin plots side by side with the spatial clustering plot\n",
    "plt.rcParams.update({'axes.titlesize': 'small'})\n",
    "\n",
    "n_types = 0\n",
    "n_types_end = 6\n",
    "obs_name = 'joint_leiden_clusters'\n",
    "\n",
    "slide = adata_slides[sample]\n",
    "\n",
    "# Subset to only plot the cell types with highest mean prevalence\n",
    "cell_types = slide.obs.iloc[:,16:].mean(axis=0).sort_values(ascending=False)[n_types:n_types_end].index\n",
    "\n",
    "fig = plt.figure(figsize=(6,14))\n",
    "\n",
    "gs = fig.add_gridspec(6,1)\n",
    "# Plot the data in the left column subplots\n",
    "for i in range(6):\n",
    "\n",
    "    ax_plotting = fig.add_subplot(gs[i,0])\n",
    "\n",
    "    sc.pl.violin(slide,groupby=obs_name,keys=cell_types[i],jitter=False,rotation=45,inner='box',\n",
    "                ax=ax_plotting,show=False)\n",
    "    \n",
    "    ax_plotting.yaxis.label.set_size(12)\n",
    "    #ax_plotting.yaxis.label.set_rotation('horizontal')\n",
    "\n",
    "    if i < 5:  # only remove x-labels for the top 5 subplots on the left\n",
    "        ax_plotting.tick_params(axis='x', which='both', bottom=False, labelbottom=False) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cropping coordinate to get a standard crop\n",
    "coords = {'PC_7875OIK':(2000,19500,2000,19500),\n",
    "          'PC_4980':(500,21000,500,21000),\n",
    "          'PC_01_14451_OIK':(1400,19000,5500,23000),\n",
    "          'PC_02_10136_VAS':(3700,22000,2200,20000),\n",
    "          'PC_02_05601_OIK':(1500,20500,3200,21000)}\n",
    "\n",
    "# Plots and saves top n_types with highest prevelance on visium slides\n",
    "sample = 'PC_02_05601_OIK'\n",
    "cell_types = ['Luminal','Luminal_Tumor','Fibroblast','T_cell']\n",
    "colormap = 'magma'\n",
    "\n",
    "slide = adata_slides[sample]\n",
    "\n",
    "# Subset to only plot the cell types with highest mean prevalence\n",
    "#cell_types = [cell_types[0]]\n",
    "\n",
    "sc.set_figure_params(dpi=600,figsize=(5,5))\n",
    "# plot in spatial coordinates\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide, cmap='viridis',\n",
    "            # show first 8 cell types\n",
    "            color=cell_types[0],\n",
    "            size=0,alpha_img=1, alpha=0,\n",
    "            # limit color scale at 99.2% quantile of cell abundance\n",
    "            vmin=0, vmax=4, show=False,colorbar_loc=None,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "            )\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/'+sample+'_histology_shown.pdf',dpi=600)\n",
    "plt.clf()\n",
    "\n",
    "for celltype in cell_types:\n",
    "\n",
    "    fig, ax = plt.subplots(1,1)\n",
    "    sc.pl.spatial(slide, cmap=colormap,\n",
    "                # show first 8 cell types\n",
    "                color=celltype,\n",
    "                size=1.4,alpha_img=0.8,\n",
    "                # limit color scale at 99.2% quantile of cell abundance\n",
    "                vmin=0, vmax=4, show=False,colorbar_loc=None,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "                )\n",
    "    plt.axis('off')\n",
    "    plt.savefig('./figures/'+sample+'_abundance_on_section_'+celltype+'.pdf',dpi=600)\n",
    "    plt.clf()\n",
    "\n",
    "\n",
    "sc.set_figure_params(dpi=600,figsize=(5,5))\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide, cmap=colormap,\n",
    "            # show first 8 cell types\n",
    "            color=cell_types[0],\n",
    "            size=1.4,alpha_img=1,\n",
    "            # limit color scale at 99.2% quantile of cell abundance\n",
    "            vmin=0, vmax=4, show=False,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "            )\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/'+sample+'_with_colorbar.pdf',dpi=600)\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# F the optimal cropping coordinates\n",
    "coords = {'PC_7875OIK':(2000,19500,2000,19500),\n",
    "          'PC_4980':(500,21000,500,21000),\n",
    "          'PC_01_14451_OIK':(1400,19000,5500,23000),\n",
    "          'PC_02_10136_VAS':(3700,22000,2200,20000),\n",
    "          'PC_02_05601_OIK':(1500,20500,3200,21000)}\n",
    "\n",
    "# Plots and saves top n_types with highest prevelance on visium slides\n",
    "sample = 'PC_02_05601_OIK'\n",
    "cell_types = ['Luminal','Luminal_Tumor','Fibroblast','T_cell']\n",
    "colormap = 'magma'\n",
    "\n",
    "slide = adata_slides[sample]\n",
    "\n",
    "# Subset to only plot the cell types with highest mean prevalence\n",
    "#cell_types = [cell_types[0]]\n",
    "\n",
    "sc.set_figure_params(dpi=100,figsize=(5,5))\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide, cmap=colormap,\n",
    "            # show first 8 cell types\n",
    "            color=cell_types[0],\n",
    "            size=1.4,alpha_img=1,\n",
    "            # limit color scale at 99.2% quantile of cell abundance\n",
    "            vmin=0, vmax=4, show=False,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "            )\n",
    "plt.axis('off')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cropping coordinate to get a standard crop\n",
    "# Cropping coordinate to get a standard crop\n",
    "coords = {'PC_7875OIK':(2000,19500,2000,19500),\n",
    "          'PC_4980':(500,21000,500,21000),\n",
    "          'PC_01_14451_OIK':(1400,19000,5500,23000),\n",
    "          'PC_02_10136_VAS':(3700,22000,2200,20000)}\n",
    "\n",
    "# Plots and saves top n_types with highest prevelance on visium slides\n",
    "sample = 'PC_02_10136_VAS'\n",
    "cell_types = ['Luminal','Luminal_Tumor','Fibroblast','T_cell']\n",
    "colormap = 'magma'\n",
    "\n",
    "slide = adata_slides[sample]\n",
    "\n",
    "# Subset to only plot the cell types with highest mean prevalence\n",
    "#cell_types = [cell_types[0]]\n",
    "\n",
    "sc.set_figure_params(dpi=600,figsize=(5,5))\n",
    "# plot in spatial coordinates\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide, cmap='viridis',\n",
    "            # show first 8 cell types\n",
    "            color=cell_types[0],\n",
    "            size=0,alpha_img=1, alpha=0,\n",
    "            # limit color scale at 99.2% quantile of cell abundance\n",
    "            vmin=0, vmax=4, show=False,colorbar_loc=None,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "            )\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/'+sample+'_histology_shown.pdf',dpi=600)\n",
    "plt.clf()\n",
    "\n",
    "for celltype in cell_types:\n",
    "\n",
    "    fig, ax = plt.subplots(1,1)\n",
    "    sc.pl.spatial(slide, cmap=colormap,\n",
    "                # show first 8 cell types\n",
    "                color=celltype,\n",
    "                size=1.4,alpha_img=0.8,\n",
    "                # limit color scale at 99.2% quantile of cell abundance\n",
    "                vmin=0, vmax=4, show=False,colorbar_loc=None,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "                )\n",
    "    plt.axis('off')\n",
    "    plt.savefig('./figures/'+sample+'_abundance_on_section_'+celltype+'.pdf',dpi=600)\n",
    "    plt.clf()\n",
    "\n",
    "\n",
    "sc.set_figure_params(dpi=600,figsize=(5,5))\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide, cmap=colormap,\n",
    "            # show first 8 cell types\n",
    "            color=cell_types[0],\n",
    "            size=1.4,alpha_img=1,\n",
    "            # limit color scale at 99.2% quantile of cell abundance\n",
    "            vmin=0, vmax=4, show=False,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "            )\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/'+sample+'_with_colorbar.pdf',dpi=600)\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plots and saves top n_types with highest prevelance on visium slides\n",
    "n_types = 12\n",
    "\n",
    "for sample in samples:\n",
    "    slide = adata_slides[sample]\n",
    "    # Subset to only plot the cell types with highest mean prevalence\n",
    "    cell_types = slide.obs.iloc[:,16:].mean(axis=0).sort_values(ascending=False)[:n_types]\n",
    "\n",
    "    # plot in spatial coordinates\n",
    "    with mpl.rc_context({'axes.facecolor':  'black',\n",
    "                        'figure.figsize': [4.5, 5]}):\n",
    "\n",
    "        sc.pl.spatial(slide, cmap='magma',\n",
    "                    # show first 8 cell types\n",
    "                    color=cell_types.index,\n",
    "                    ncols=4, size=1.3,alpha_img=0.8,\n",
    "                    # limit color scale at 99.2% quantile of cell abundance\n",
    "                    vmin=0, vmax='p99.2', show=False\n",
    "                    )\n",
    "        plt.savefig('./plots/c2l_mapping_results_20230322/'+sample+'_c2l_mapping_top12_abundant.png',dpi=200)\n",
    "        plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Side by side violin plots of abundant cell types + spatial \n",
    "\n",
    "plt.rcParams.update({'axes.titlesize': 'small'})\n",
    "\n",
    "n_types = 6\n",
    "obs_name = 'joint_leiden_clusters'\n",
    "\n",
    "for s in samples:\n",
    "\n",
    "    slide = adata_slides[s]\n",
    "\n",
    "    # Subset to only plot the cell types with highest mean prevalence\n",
    "    cell_types = slide.obs.iloc[:,16:].mean(axis=0).sort_values(ascending=False)[:n_types].index\n",
    "\n",
    "    fig = plt.figure(figsize=(12,12))\n",
    "\n",
    "    gs = fig.add_gridspec(6,3)\n",
    "    ax6 = fig.add_subplot(gs[:, 1:3])\n",
    "\n",
    "    # Plot the data in the left column subplots\n",
    "    for i in range(6):\n",
    "\n",
    "        ax_plotting = fig.add_subplot(gs[i,0])\n",
    "\n",
    "        sc.pl.violin(slide,groupby=obs_name,keys=cell_types[i],jitter=False,rotation=45,inner='box',\n",
    "                    ax=ax_plotting,show=False)\n",
    "        \n",
    "        ax_plotting.yaxis.label.set_size(12)\n",
    "        #ax_plotting.yaxis.label.set_rotation('horizontal')\n",
    "\n",
    "        if i < 5:  # only remove x-labels for the top 5 subplots on the left\n",
    "            ax_plotting.tick_params(axis='x', which='both', bottom=False, labelbottom=False) \n",
    "\n",
    "    sc.pl.spatial(adata_slides[s],color='joint_leiden_clusters',size=1.3,ax=ax6)\n",
    "\n",
    "#    plt.show()\n",
    "    fig.savefig('./plots/c2l_cell_type_prevalence_violin_plots/'+s+'_cell_type_prevalences_per_cluster.png')\n",
    "    fig.clf()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cell2loc_adata = ad.AnnData(adata_vis.obs.iloc[:,19:])\n",
    "cell2loc_adata.obs['sample'] = adata_vis.obs['sample_id']\n",
    "cell2loc_adata.obs['clusters'] = adata_vis.obs['sample_id'].astype('str') + '_' +  adata_vis.obs['joint_leiden_clusters'].astype('str')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scoring gene markers on spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "signature_dict = {\n",
    "# Full signature from supplemntary table 3 of https://doi.org/10.1038/s41467-021-25624-1\n",
    "'Persist' : ['CDC20','MKI67','CCNB2','CENPF','DLGAP5','HMMR','PLK1','CCNB1','PTTG1','CENPE','UBE2S','NCAPD2','TUBA1C','DYNLL1','CD81','CKAP5','ODC1'],\n",
    "\n",
    "# Top 50 genes from the first tab of supplementary table 1 in https://doi.org/10.1016/j.ccell.2019.06.005\n",
    "'Balanis_NE' : ['PIK3R3','TARDBP','TMBIM4','DPYSL5','INSM1','PIK3R2','TIMM10B','PSMA1','DLL3','ASCL1','FOXG1','ZIC2','SOX11','ZIC5','LHX2','DLK1',\n",
    "              'PI4K2B','SEZ6','ACTL6B','PCSK1','CACNA1A','PAPD7','CPLX2','XKR7','ZNF286A','ST18','INA','CELF3','TAGLN3','PAGR1','NRSN1','ONECUT2',\n",
    "              'ELAVL3','POU4F1','CDK5R2','SCG3','DCX','BARX1','FAM57B','C17orf100','UNC13A','CHRNB2','ST8SIA3','NEUROD1','MAST1','ELAVL4','KIF1A','ACSL6','SPATA13','KIF18B'],\n",
    "\n",
    "# From supplementary table 1 of https://doi.org/10.1016/j.clgc.2020.08.004\n",
    "'TP53_1' :['ABR','ACTL6A','ADAM17','AK2','ALDH3A2','ANKRA2','ANLN','ASF1B','ASPM','ATP9B','AURKA','BBS4','BEND3','BRIX1','BUB1','C16orf71','C17orf53','C17orf59','C9orf117',\n",
    "          'CCDC150','CCDC40','CCNA2','CCNB2','CCNG1','CDC123','CDC20','CDC25C','CDC45','CDCA4','CDCA5','CDCA8','CDKN1A','CDKN3','CDON','CDR2L','CENPA','CENPF','CENPI','CENPL',\n",
    "          'CES2','CIT','CKAP2L','CPSF3','CSNK2A1','CYB5D1','CYB5D2','DCAF13','DCUN1D5','DDB2','DEPDC1B','DERL1','DGAT2','DLGAP5','DNMT3B','DPH2','DSN1','DVL3','E2F2','E2F3','ECT2'\n",
    "          ,'EDA2R','EEF2','ERCC6L','ESPL1','EXO1','EZH2','FAM49B','FAM72B','FANCB','FANCF','FDXR','FMO5','GINS1','GINS4','GNAI3','GREB1','GTF3C2','GTPBP4','HHAT','HJURP','INCENP',\n",
    "          'IQGAP3','ISY1','ITGB1BP1','ITGBL1','KCNK6','KIAA1524','KIF18B','KIF20A','KIF23','KIF2C','KIF4A','KIF4B','KIFC1','KPNA2','LSG1','MAP2K4','MCM10','MCM4','MDM2','MED11','MELK',\n",
    "          'MLF1','MPDU1','MRPL13','MRPL3','MSH2','MTBP','MYBL2','NAA50','NCAPG2','NCAPG','NCAPH','NCBP2','NEIL3','NFATC3','NUF2','NUP85','PARP2','PDCD10','PHLDA3','PKMYT1','PLK1','POLQ',\n",
    "          'POLR2K','PRPF38A','PSMA7','PSMC2','PSMD12','PTTG1','RAB6B','RABEP1','RACGAP1','RAD21','RAD51','RAD54B','RAD54L','RAE1','RANGRF','RBM17','RBM28','RFC4','RPH3AL','RPS27L','RSPH1',\n",
    "          'SEC22A','SGOL1','SMC1B','SMC4','SMG6','SMYD2','SNRPB2','SNX33','SPAG5','SPATA18','SPDEF','SPRY3','SPTBN4','SRR','STIP1','SYTL1','TACC3','TARDBP','TCEB1','TK1','TMEM132A','TNFSF12',\n",
    "          'TOPBP1','TPX2','TRAIP','TRIP13','TROAP','TTC9C','TTK','UBE2C','VAMP2','VPS39','VRK1','VWA3A','WDHD1','WDR53','WDR62','WDR78','XPO1','YEATS2','ZNF18'],\n",
    "\n",
    "\n",
    "# These below are all from supplementary table 16 of https://doi.org/10.1126/science.abe1505\n",
    "'TCGA_androgen' : ['PTGER4','FKBP5','KLK2','CENPN','MAF','ACSL3','HERC3','ZBTB10','EAF2','ABCC4','C1orf116','PMEPA1','MED28','MPHOSPH9','TMPRSS2','KLK3','NKX3-1','NNMT','ADAM7','ELL2'],\n",
    "'Beltran_NE' : ['ASXL3','CAND2','ETV5','GPX2','JAKMIP2','KIAA0408','SOGA3','TRIM9','BRINP1','C7orf76','GNAO1','KCNB2','KCND2','LRRC16B','MAP10','NRSN1','PCSK1','PROX1','RGS7','SCG3','SEC11C','SEZ6','ST8SIA3','SVOP','SYT11','AURKA','DNMT1','EZH2','MYCN'],\n",
    "'WNT_hallmark' : ['ADAM17','AXIN1','AXIN2','CCND2','CSNK1E','CTNNB1','CUL1','DKK1','DKK4','DLL1','DVL2','FRAT1','FZD1','FZD8','GNAI1','HDAC11','HDAC2','HDAC5','HEY1','HEY2','JAG1','JAG2','KAT2A','LEF1',\n",
    "                'MAML1','MYC','NCOR2','NCSTN','NKD1','NOTCH1','NOTCH4','NUMB','PPARD','PSEN2','PTCH1','RBPJ','SKP2','TCF7','TP53','WNT1','WNT5B','WNT6'],\n",
    "'Wang_YAP_TAZ' : ['CYR61','CTGF','AMOTL2','ANKRD1','IGFBP3','F3','FJX1','NUAK2','LAST2','CRIM1','GADD45A','TGFB2','PTPN14','NTSE','FOXF2','AXL','DOCK5','ASAP1','RBMS3','MYOF','ANHGEF17','CCDC80','AJUBA']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate gene scores for both aggregated visium data and individual slides (the same calculation)\n",
    "for k in signature_dict.keys():\n",
    "    \n",
    "    sc.tl.score_genes(adata=adata_vis,gene_list=signature_dict[k],score_name=k+'_score')\n",
    "\n",
    "    for s in samples:\n",
    "        sc.tl.score_genes(adata=adata_slides[s],gene_list=signature_dict[k],score_name=k+'_score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pseudobulk violin plots of gene set scores\n",
    "sc.set_figure_params(figsize=(12,6))\n",
    "for k in signature_dict.keys():\n",
    "    fig,ax = plt.subplots(1,1)\n",
    "    sc.pl.violin(adata_vis,groupby='sample_id',keys=k+'_score',legend_loc=None,inner='box',jitter=False,ax=ax,show=False,rotation=60)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('./plots/gene_score_violin_plots/'+k+'_violin_plots_pseudobulk.png',dpi=120)\n",
    "    plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Side by side violin plots of gene set scores + spatial \n",
    "\n",
    "plt.rcParams.update({'axes.titlesize': 'small'})\n",
    "\n",
    "n_types = 7\n",
    "obs_name = 'joint_leiden_clusters'\n",
    "\n",
    "for s in samples:\n",
    "\n",
    "    adata_subset = adata_slides[s]\n",
    "\n",
    "    fig = plt.figure(figsize=(16,16))\n",
    "\n",
    "    gs = fig.add_gridspec(7,3)\n",
    "    ax6 = fig.add_subplot(gs[:, 1:3])\n",
    "\n",
    "    # Plot the data in the left column subplots\n",
    "    for i,k in enumerate(signature_dict.keys()):\n",
    "\n",
    "        ax_plotting = fig.add_subplot(gs[i,0])\n",
    "\n",
    "        sc.tl.score_genes(adata=adata_subset,gene_list=signature_dict[k],score_name=k+'_score')\n",
    "        sc.pl.violin(adata_subset,groupby='joint_leiden_clusters',keys=k+'_score',legend_loc=None,inner='box',jitter=False,ax=ax_plotting,show=False)\n",
    "        \n",
    "        ax_plotting.yaxis.label.set_size(12)\n",
    "        #ax_plotting.yaxis.label.set_rotation('horizontal')\n",
    "\n",
    "        if i < 6:  # only remove x-labels for the top 5 subplots on the left\n",
    "            ax_plotting.tick_params(axis='x', which='both', bottom=False, labelbottom=False) \n",
    "\n",
    "    sc.pl.spatial(adata_subset,color='joint_leiden_clusters',size=1.3,ax=ax6,show=False)\n",
    "\n",
    "#    plt.show()\n",
    "    plt.savefig('./plots/gene_score_violin_plots/'+s+'_gene_set_scores.png',dpi=120)\n",
    "    plt.clf()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main differential expression test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PC_samples = samples[3:13]\n",
    "col_name = 'joint_leiden_clusters'\n",
    "\n",
    "\n",
    "de_res = pd.DataFrame()\n",
    "for sample in PC_samples:\n",
    "    slide = adata_slides[sample]\n",
    "    groups = list(slide.obs[col_name].cat.categories)\n",
    "\n",
    "    sc.tl.rank_genes_groups(slide, col_name, method='wilcoxon') # groups=['1','2'],reference='0',\n",
    "    res = sc.get.rank_genes_groups_df(slide,groups)\n",
    "    res = res[(res['pvals_adj']<0.05) & (res['logfoldchanges']>=1)].sort_values('logfoldchanges',ascending=False).reset_index(drop=True)\n",
    "\n",
    "    res['group'] = sample + '_' + res['group'].astype(str)\n",
    "    de_res = pd.concat([de_res,res],axis=0)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take out genes that are in only one cluster\n",
    "\n",
    "# The minimum number can be modified if \n",
    "de_genes_for_use = list(de_res['names'].value_counts()[de_res['names'].value_counts() > 1].keys()) \n",
    "\n",
    "len(de_genes_for_use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset PC samples and calculate scaled expression (Z-scores)\n",
    "adata_pc = adata_vis[adata_vis.obs['sample_id'].isin(PC_samples)].copy()\n",
    "adata_pc.layers['scaled'] = sc.pp.scale(adata_pc, copy=True,layer='counts').X\n",
    "\n",
    "\n",
    "# Subset the adata object by calculating an overlap\n",
    "adata_pc = adata_pc[:,adata_pc.var_names.isin(de_genes_for_use)]\n",
    "de_genes_for_use = [g for g in de_genes_for_use if g in adata_pc.var_names]\n",
    "adata_pc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrich_results(adata,cluster):\n",
    "    df = sc.get.rank_genes_groups_df(adata,group=cluster)\n",
    "    df = df[(df['pvals_adj']<0.05) & (df['logfoldchanges']>=1)].sort_values('logfoldchanges',ascending=False)\n",
    "    enriched = sc.queries.enrich(list(df['names']),gprofiler_kwargs={'sources':['GO:BP']})\n",
    "    return(enriched)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'PC_02_05601_OIK'\n",
    "groups = ['0','1','2','3','4','5','6','7']\n",
    "\n",
    "slide = adata_slides[sample].copy()\n",
    "slide.layers['scaled'] = sc.pp.scale(slide, copy=True,layer='counts').X\n",
    "slide = slide[slide.obs['joint_leiden_clusters'].isin(groups)]\n",
    "\n",
    "sc.tl.rank_genes_groups(slide, \"joint_leiden_clusters\", method='wilcoxon') # groups=['1','2'],reference='0',"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res = pd.DataFrame()\n",
    "for group in groups:\n",
    "    res = sc.get.rank_genes_groups_df(slide,group)\n",
    "    res[(res['pvals_adj']<0.05) & (res['logfoldchanges']>=1)].sort_values('logfoldchanges',ascending=False)\n",
    "\n",
    "    res['group'] = group\n",
    "\n",
    "    res = res.iloc[:5]\n",
    "    de_res = pd.concat([de_res,res],axis=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_markers = {}\n",
    "for group in groups:\n",
    "    res = sc.get.rank_genes_groups_df(slide,group)\n",
    "    res[(res['pvals_adj']<0.05) & (res['logfoldchanges']>=1)].sort_values('logfoldchanges',ascending=False)\n",
    "    \n",
    "    de_markers[group] = res.loc[:50]['names']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gr in de_markers.keys():\n",
    "    markers = de_markers[gr]\n",
    "    for marker in markers:\n",
    "        for sign in signature_dict.keys():\n",
    "            if marker in signature_dict[sign]:\n",
    "                print(sign+' gene '+marker+' in cluster ' + gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(slide, de_markers,cmap='RdBu_r', groupby='joint_leiden_clusters', layer='scaled', swap_axes=True,dendrogram=True,vmin=-2,vmax=2,\n",
    "                   figsize=(10,10)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(slide, de_markers,cmap='RdBu_r', groupby='joint_leiden_clusters', layer='scaled', swap_axes=True,dendrogram=True,vmin=-2,vmax=2,\n",
    "                   figsize=(8,8),save='_'+sample+'_expression.pdf'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrich_res = pd.DataFrame()\n",
    "for group in groups:\n",
    "    res = enrich_results(slide,group)\n",
    "    res['group'] = group\n",
    "    enrich_res = pd.concat([enrich_res,res],axis=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Heatmap plotting of enrichment results for PC_02_10136\n",
    "\n",
    "tokeep = ['mesenchyme morphogenesis','vasculature development','response to stress','acyl-CoA biosynthetic process']\n",
    "res_for_plotting = enrich_res[enrich_res['name'].isin(tokeep)]\n",
    "\n",
    "# Format the results for plotting\n",
    "res_for_plotting = res_for_plotting[['name','p_value','intersection_size','group']]\n",
    "x_coords ={'0':5,'1':10,'2':15,'5':20}\n",
    "y_coords ={'mesenchyme morphogenesis':5,'response to stress':10,'acyl-CoA biosynthetic process':15,'vasculature development':20}\n",
    "res_for_plotting['x'] = res_for_plotting['group'].map(x_coords)\n",
    "res_for_plotting['y'] = res_for_plotting['name'].map(y_coords)\n",
    "res_for_plotting['minus_log10_p_val'] = -np.log(res_for_plotting['p_value'])\n",
    "res_for_plotting = res_for_plotting.reset_index(drop=True)\n",
    "res_for_plotting\n",
    "\n",
    "df_heatmap = pd.pivot(res_for_plotting,index='name',values=['minus_log10_p_val'],columns='group')\n",
    "df_heatmap = df_heatmap.fillna(0)\n",
    "df_heatmap = df_heatmap.iloc[[1,2,0,3]] # Reorder to get a diagonal\n",
    "\n",
    "sns.set(rc={'figure.figsize':(4,2),\"figure.dpi\": 300})\n",
    "fig,ax = plt.subplots(1,1)\n",
    "ax = sns.heatmap(df_heatmap.T,vmax=20, cmap='Blues', linewidths=1.5, linecolor='white', cbar=None\n",
    "                )\n",
    "ax.axis('off')\n",
    "plt.savefig('figures/enrichment_heatmap_no_axis.pdf',bbox_inches='tight',dpi=300)\n",
    "df_heatmap = pd.pivot(res_for_plotting,index='name',values=['minus_log10_p_val'],columns='group')\n",
    "df_heatmap = df_heatmap.fillna(0)\n",
    "df_heatmap = df_heatmap.iloc[[1,2,0,3]] # Reorder to get a diagonal\n",
    "\n",
    "sns.set(rc={'figure.figsize':(4,2),\"figure.dpi\": 300})\n",
    "fig,ax = plt.subplots(1,1)\n",
    "ax = sns.heatmap(df_heatmap.T,vmax=20, cmap='Blues'#,cbar=None\n",
    "                )\n",
    "#ax.axis('off')\n",
    "plt.savefig('figures/enrichment_heatmap_with_axis.pdf',bbox_inches='tight',dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={'figure.figsize':(4,2),\"figure.dpi\": 300})\n",
    "fig,ax = plt.subplots(1,1)\n",
    "\n",
    "ax = sns.scatterplot(data=res_for_plotting, x=\"x\", y=\"y\", hue='minus_log10_p_val', size=\"intersection_size\",\n",
    "    palette='viridis' , legend=None,sizes=(50, 500), hue_norm=(5, 60)\n",
    ")\n",
    "#sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "ax.grid(False)\n",
    "ax.axis('off')\n",
    "ax.set_ylim([25,3])\n",
    "ax.set_xlim([3,25])\n",
    "#ax.invert_yaxis()\n",
    "plt.tight_layout()\n",
    "plt.savefig('figures/enrichment_scatterplot_no_axis.pdf',bbox_inches='tight',dpi=300)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting accurately cropped visium slides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cropping coordinate to get a standard crop\n",
    "coords = {'PC_7875OIK':(2000,19500,2000,19500),\n",
    "          'PC_4980':(500,21000,500,21000),\n",
    "          'PC_01_14451_OIK':(1400,19000,5500,23000),\n",
    "          'PC_02_10136_VAS':(3700,22000,2200,20000)}\n",
    "\n",
    "# Plots and saves top n_types with highest prevelance on visium slides\n",
    "sample = 'PC_02_10136_VAS'\n",
    "slide = adata_slides[sample]\n",
    "\n",
    "# Subset to only plot the cell types with highest mean prevalence\n",
    "#cell_types = [cell_types[0]]\n",
    "\n",
    "sc.set_figure_params(dpi=100,figsize=(6,6))\n",
    "# plot in spatial coordinates\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide,\n",
    "            color='joint_leiden_clusters',\n",
    "            size=1.4,alpha_img=0.8,\n",
    "            show=False,ax = ax,crop_coord=coords[sample]\n",
    "            )\n",
    "plt.axis('off')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cropping coordinate to get a standard crop\n",
    "coords = {'PC_7875OIK':(2000,19500,2000,19500),\n",
    "          'PC_4980':(500,21000,500,21000),\n",
    "          'PC_01_14451_OIK':(1400,19000,5500,23000),\n",
    "          'PC_02_10136_VAS':(3700,22000,2200,20000),\n",
    "          'PC_02_05601_OIK':(1500,20500,3200,21000),\n",
    "          'CRPC-278':(1500,20500,3200,21000),\n",
    "          'BPH_665':(2800,21500,2200,20500),\n",
    "          'PC_03_01669_TUTKV':(0,18000,3000,21000),\n",
    "          'PC-03-6712':(0,18000,3000,21000)\n",
    "          }\n",
    "\n",
    "# Plots and saves top n_types with highest prevelance on visium slides\n",
    "sample = 'PC-03-6712'\n",
    "slide = adata_slides[sample]\n",
    "\n",
    "# Subset to only plot the cell types with highest mean prevalence\n",
    "#cell_types = [cell_types[0]]\n",
    "\n",
    "sc.set_figure_params(dpi=100,figsize=(3,3))\n",
    "# plot in spatial coordinates\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide,\n",
    "            color='joint_leiden_clusters',\n",
    "            size=0,alpha_img=1, alpha=0,\n",
    "            # limit color scale at 99.2% quantile of cell abundance\n",
    "            show=False,ax = ax,crop_coord=coords[sample]  # 'p99.2'\n",
    "            )\n",
    "\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/'+sample+'_histology_shown.pdf',dpi=800)\n",
    "plt.clf()\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "sc.pl.spatial(slide,\n",
    "            color='joint_leiden_clusters',\n",
    "            size=1.4,alpha_img=0.8,\n",
    "            show=False,ax = ax,crop_coord=coords[sample]\n",
    "            )\n",
    "plt.axis('off')\n",
    "plt.savefig('./figures/'+sample+'_clustering.pdf',dpi=800)\n",
    "plt.clf()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
