{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Date: 9.8.2023\n",
    "# Author: Antti Kiviaho\n",
    "#\n",
    "# This notebook handles clustered Visium data with the goal of\n",
    "# analyzing ligand-receptor interactions between\n",
    "# 1. Joint leiden clusters (spatial proximity + gene expression)\n",
    "# 2. A cluster of interest and it's immediate surroundings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "os.chdir('/lustre/scratch/kiviaho/prostate_spatial/')\n",
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import squidpy as sq\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from scripts.utils import load_from_pickle, get_sample_ids, save_to_pickle, get_treatment_info\n",
    "import matplotlib as mpl\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_theme()\n",
    "\n",
    "sc.set_figure_params(figsize=(8,8))\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "\n",
    "# import matplotlib library\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from scipy.stats import ttest_rel\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Define functions\n",
    "def get_annotation_proximity_spots(dat, cluster_of_interest, interaction_cluster, added_key='proximity_analysis'):\n",
    "\n",
    "    ''' This function is used when clusters have been annotated as one of three:\n",
    "        1. tumor\n",
    "        2. normal\n",
    "        3. stroma\n",
    "    and must be saved in 'annotation' observation column\n",
    "    See above for the procedure of creating these annotations.'''\n",
    "\n",
    "    # Create an observation column for spatial segmentation\n",
    "    dat.obs[added_key] = 'background'\n",
    "    distance_mat = dat.obsp['spatial_distances'].todense()\n",
    "\n",
    "    for idx, obs_name in enumerate(dat.obs_names):\n",
    "        cl = dat.obs['annotation'][idx]\n",
    "\n",
    "        if cl == cluster_of_interest:\n",
    "\n",
    "            first_nhbor_idxs = np.where(distance_mat[:,idx]==1.0)[0] # Get first-term neighbor indices\n",
    "\n",
    "            try:\n",
    "                # If try fails, there are no matching clusters as keys in value_counts\n",
    "                n_cl_neighbors = dat[first_nhbor_idxs].obs['annotation'].value_counts()[cl] # find first-term neighbor cluster annotations POSSIBLE ERROR IF CL NOT IN DICT\n",
    "\n",
    "                # Added this clause to control that only those with 'close' interactions with the interaction cluster are included. \n",
    "                all_nhbor_indices = np.where(distance_mat[:, idx] != 0)[0]\n",
    "\n",
    "                # Downgraded the number of required first neighbors to two\n",
    "                if (n_cl_neighbors >= 2) & (sum(dat.obs['annotation'][all_nhbor_indices] == interaction_cluster) >= 3):\n",
    "                    dat.obs.at[obs_name,added_key] = cl\n",
    "\n",
    "            except:\n",
    "                continue\n",
    "\n",
    "    # Make a second loop to make sure the final cluster-of-interest annotations\n",
    "    # are what's used to define proximal spots\n",
    "    for idx, obs_name in enumerate(dat.obs_names):\n",
    "        cl = dat.obs[added_key][idx]\n",
    "\n",
    "        if cl == cluster_of_interest:\n",
    "            \n",
    "            all_nhbor_indices = np.where(distance_mat[:, idx] != 0)[0] \n",
    "\n",
    "            # Get the indices where neighboring spots are not the interest cluster \n",
    "            indices = np.where((dat.obs[added_key][all_nhbor_indices] != cl) & (dat.obs['annotation'][all_nhbor_indices] == interaction_cluster))[0]\n",
    "\n",
    "            # Update the 'proximity_analysis' column for the specific indices\n",
    "            dat.obs.loc[dat.obs_names[all_nhbor_indices[indices]], added_key] = 'proximal_' + interaction_cluster\n",
    "\n",
    "    # Modify the colors to maintain the original cluster color\n",
    "    dat.obs[added_key] = dat.obs[added_key].astype('category')\n",
    "\n",
    "    return(dat)\n",
    "\n",
    "def most_common_interactions(dct):\n",
    "    tuple_lst = []\n",
    "    for key in dct.keys():\n",
    "        [tuple_lst.append(tup) for tup in dct[key]]\n",
    "    res = pd.Series(tuple_lst).value_counts()\n",
    "    return(res)\n",
    "\n",
    "def merge_dicts(dict1, dict2):\n",
    "    merged_dict = {}\n",
    "\n",
    "    for key, value in dict1.items():\n",
    "        if key in dict2:\n",
    "            merged_dict[key] = [value, dict2[key], value - dict2[key]]\n",
    "        else:\n",
    "            merged_dict[key] = [value, 0, value]\n",
    "\n",
    "    for key, value in dict2.items():\n",
    "        if key not in merged_dict:\n",
    "            merged_dict[key] = [0, value, 0 - value]\n",
    "\n",
    "    return merged_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_slides = load_from_pickle('./data/clustered_visium_data.pickle')\n",
    "samples = get_sample_ids()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instead of manually selecting interactions, make a script that automatically finds the highest mean cell type\n",
    "# and if it is tumor, annotate it as such (or fibroblast, luminal, basal etc.)\n",
    "cell_mapping_dat = sc.read_h5ad('c2l_mapping_as_anndata_20230721.h5ad')\n",
    "\n",
    "\n",
    "# Normalize inferred cell counts to cell-type proportions\n",
    "\n",
    "arr = cell_mapping_dat.X.copy()\n",
    "# Calculate the row sum\n",
    "row_sum = np.sum(arr, axis=1)\n",
    "\n",
    "# Calculate the scaling factor for each row\n",
    "scaling_factor = 100 / row_sum\n",
    "\n",
    "# Multiply each row with its respective scaling factor\n",
    "scaled_arr = np.multiply(arr.T, scaling_factor).T\n",
    "\n",
    "cell_mapping_dat.X = scaled_arr.copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "representative_celltypes = {}\n",
    "n_max_cell_comb= 2\n",
    "\n",
    "# Initialize the dict in case top number of celltypes does not satisfy the p-value threshold\n",
    "for c in cell_mapping_dat.obs['joint_leiden_clusters'].cat.categories:\n",
    "    representative_celltypes[c] = [None,np.nan]\n",
    "\n",
    "for cluster in cell_mapping_dat.obs['joint_leiden_clusters'].cat.categories:\n",
    "    subset = cell_mapping_dat[cell_mapping_dat.obs['joint_leiden_clusters'] == cluster].copy()\n",
    "\n",
    "    for split_idx in (np.arange(n_max_cell_comb)+1)[::-1]:\n",
    "        idxs = np.argsort(np.round(np.mean(subset.X,axis=0)))[::-1].tolist()\n",
    "        top_5 = list(subset.var_names[idxs])[:5]\n",
    "        top_ctypes = list(subset.var_names[idxs])[:split_idx]\n",
    "        rest_ctypes = list(subset.var_names[idxs])[split_idx:]\n",
    "\n",
    "        #sc.pl.violin(subset,keys=top_5,jitter=False,rotation=30,inner='box',palette='Set2')\n",
    "\n",
    "        pval = ttest_rel(\n",
    "            np.sum(subset[:,top_ctypes].X.copy(),axis=1).ravel(),\n",
    "            np.sum(subset[:,rest_ctypes].X.copy(),axis=1).ravel(),alternative='greater')[1]\n",
    "\n",
    "        if pval < 0.05:\n",
    "            if split_idx > 1:\n",
    "                representative_celltypes[cluster] = ['_'.join(top_ctypes),pval]\n",
    "            else:\n",
    "                representative_celltypes[cluster] = [top_ctypes[0],pval]\n",
    "\n",
    "\n",
    "res_df = pd.DataFrame(representative_celltypes,index=['cell-type combination','p-value']).T\n",
    "\n",
    "\n",
    "# Get the p-values from res_df in a separate variable\n",
    "p_values = res_df['p-value']\n",
    "\n",
    "# Perform the Benjamini-Hochberg correction\n",
    "corrected_values = multipletests(p_values, method='fdr_bh')\n",
    "\n",
    "# Update the 'p-value' column in res_df with the corrected values\n",
    "res_df['corr_p-value'] = corrected_values[1]\n",
    "\n",
    "res_df = res_df[res_df['corr_p-value']<0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "stroma    63\n",
       "tumor     62\n",
       "normal    39\n",
       "Name: cell-type combination, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Annotate clusters to an uniform format\n",
    "res_df[(res_df == 'myofibroblasts_inflammatory fibroblasts') | (res_df == 'inflammatory fibroblasts_myofibroblasts') | (res_df == 'myofibroblasts') | (res_df== 'inflammatory fibroblasts') ] = 'stroma'\n",
    "res_df[(res_df == 'ERG tumor epithelium') | (res_df == 'luminal epithelium_ERG tumor epithelium') | (res_df == 'ERG tumor epithelium_luminal epithelium') | (res_df == 'ERG tumor epithelium_inflammatory fibroblasts') | (res_df == 'ERG tumor epithelium_myofibroblasts')] = 'tumor'\n",
    "res_df[(res_df == 'luminal epithelium') | (res_df == 'luminal epithelium_inflammatory fibroblasts') | (res_df == 'luminal epithelium_normal basal epithelium') | (res_df == 'luminal epithelium_myofibroblasts')] = 'normal'\n",
    "\n",
    "# Drop rows with uniqe cell type combinations\n",
    "res_df = res_df[~res_df['cell-type combination'].isin(list(res_df['cell-type combination'].value_counts().index[res_df['cell-type combination'].value_counts() <2]))]\n",
    "\n",
    "res_df['cell-type combination'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the annotation column into each member of the adata_slides object\n",
    "cluster_annotation_dict = res_df['cell-type combination'].to_dict()\n",
    "\n",
    "for sample in samples:\n",
    "    dat = adata_slides[sample].obs.copy()\n",
    "    dat['sample_cluster'] = dat['sample_id'].astype(str) + '_' + dat['joint_leiden_clusters'].astype(str)\n",
    "    adata_slides[sample].obs['annotation'] = dat['sample_cluster'].map(cluster_annotation_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors_dict = {\n",
    "    'tumor_normal':['#919191','#74b572','#db1616'],\n",
    "    'tumor_stroma':['#919191','#f7f774','#db1616'],\n",
    "    'normal_tumor':['#919191','#74b572','#db1616'],\n",
    "    'normal_stroma':['#919191','#74b572','#f7f774'],\n",
    "    'stroma_normal':['#919191','#74b572','#f7f774'],\n",
    "    'stroma_tumor':['#919191','#db1616','#f7f774']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 600x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "source = 'tumor'\n",
    "target = 'normal'\n",
    "cols = colors_dict['tumor_normal']\n",
    "interaction_dict = {}\n",
    "\n",
    "for sample in samples:\n",
    "\n",
    "    slide = adata_slides[sample].copy()\n",
    "\n",
    "    if (slide.obs['annotation'].str.contains(source).any()) & (slide.obs['annotation'].str.contains(target).any()):\n",
    "        # Using three rings you get 6 immediate neigbors, 12 second neigbors and 18 third neighbors\n",
    "        sq.gr.spatial_neighbors(slide,n_neighs=6,n_rings=4)\n",
    "\n",
    "        slide = get_annotation_proximity_spots(slide, source, target)\n",
    "    \n",
    "        sq.gr.ligrec(\n",
    "            slide,\n",
    "            n_perms=100,\n",
    "            cluster_key=\"proximity_analysis\",\n",
    "            show_progress_bar = False\n",
    "        )\n",
    "\n",
    "        proximal_spots = 'proximal_' + target\n",
    "\n",
    "        pvals = slide.uns['proximity_analysis_ligrec']['pvalues'][source][proximal_spots]\n",
    "        means = slide.uns['proximity_analysis_ligrec']['means'][source][proximal_spots]\n",
    "        tuple_array = pd.DataFrame(means[pvals<0.01][means>1]).index.values\n",
    "\n",
    "        interaction_dict[sample] = tuple_array\n",
    "\n",
    "        ## Plotting ##\n",
    "        slide.uns['proximity_analysis_colors'] = cols\n",
    "\n",
    "        # set figure axis size and dpi\n",
    "        fig, ax = plt.subplots(figsize=(5, 5), dpi=120)\n",
    "\n",
    "        # create spatial plot\n",
    "        sc.pl.spatial(slide,color='proximity_analysis',size=1.3,alpha=0.8, ax = ax, show= False, title= sample)\n",
    "        plt.tight_layout()\n",
    "\n",
    "        # create filename with sample name\n",
    "        filename = 'plots/receptor_ligand_interaction_analysis/' + sample + '_'+ source +'_to_'+ target +'_clusters_communication.pdf'\n",
    "\n",
    "        # create output folder if it doesn't exist\n",
    "        if not os.path.exists(os.path.dirname(filename)):\n",
    "            os.makedirs(os.path.dirname(filename))\n",
    "\n",
    "        # save plot to pdf with filename\n",
    "        with PdfPages(filename) as pdf:\n",
    "            pdf.savefig(fig)\n",
    "            plt.clf()\n",
    "\n",
    "save_to_pickle(interaction_dict,'./data/'+source+'_to_'+target+'_ligand_receptor_proximity_interactions.pickle')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Performing treament-status based differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "treatment_info = get_treatment_info()\n",
    "\n",
    "source = 'tumor'\n",
    "target = 'normal'\n",
    "concat_list = []\n",
    "\n",
    "ligand_receptor_dict = load_from_pickle('data/'+source+'_to_'+target+'_ligand_receptor_proximity_interactions.pickle')\n",
    "\n",
    "treated = {}\n",
    "untreated = {}\n",
    "\n",
    "for s in ligand_receptor_dict.keys():\n",
    "\n",
    "    if (treatment_info[s] == 'goserelin') | (treatment_info[s] == 'bicalutamide'):\n",
    "        treated[s] = ligand_receptor_dict[s]\n",
    "\n",
    "    if (treatment_info[s] == 'untreated'):\n",
    "        untreated[s] = ligand_receptor_dict[s]\n",
    "        \n",
    "\n",
    "for sample in samples:\n",
    "\n",
    "    slide = adata_slides[sample].copy()\n",
    "\n",
    "    if (slide.obs['annotation'].str.contains(source).any()) & (slide.obs['annotation'].str.contains(target).any()):\n",
    "        # Using three rings you get 6 immediate neigbors, 12 second neigbors and 18 third neighbors\n",
    "        sq.gr.spatial_neighbors(slide,n_neighs=6,n_rings=4)\n",
    "\n",
    "        slide = get_annotation_proximity_spots(slide, source, target)\n",
    "\n",
    "        concat_list.append(slide)\n",
    "\n",
    "dat_concat = ad.concat(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [],
   "source": [
    "treated_most_common = most_common_interactions(treated)\n",
    "untreated_most_common = most_common_interactions(untreated)\n",
    "merged_dict = merge_dicts(untreated_most_common,treated_most_common)\n",
    "\n",
    "df = pd.DataFrame(merged_dict,index=[\n",
    "    'untreated '+ '('+str(len(untreated))+')',\n",
    "    'treated  '+ '('+str(len(treated))+')',\n",
    "    'diff']).T.sort_values('diff',ascending=False)\n",
    "\n",
    "ligands = np.array([tup[0] for tup in list(df.index)])\n",
    "_, idx = np.unique(ligands, return_index=True)\n",
    "ligands = ligands[np.sort(idx)].tolist()\n",
    "\n",
    "receptors = np.array([tup[1] for tup in list(df.index)])\n",
    "_, idx = np.unique(receptors, return_index=True)\n",
    "receptors = receptors[np.sort(idx)].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>untreated (6)</th>\n",
       "      <th>treated  (2)</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GDF15</th>\n",
       "      <th>ERBB3</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMTM3</th>\n",
       "      <th>KLK3</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">GDF15</th>\n",
       "      <th>ERBB2</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFBR2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CALR</th>\n",
       "      <th>P4HB</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAP2</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRP1</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">GDF15</th>\n",
       "      <th>EGFR</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTK2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDK</th>\n",
       "      <th>NCL</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HLA-B</th>\n",
       "      <th>KLRD1</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LGALS3</th>\n",
       "      <th>ANXA2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDK</th>\n",
       "      <th>SDC1</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HLA-B</th>\n",
       "      <th>CD8B</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPA8</th>\n",
       "      <th>MAPK9</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">VEGFA</th>\n",
       "      <th>ERBB2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFE2L2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HLA-B</th>\n",
       "      <th>CD3G</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               untreated (6)  treated  (2)  diff\n",
       "GDF15  ERBB3               5             0     5\n",
       "CMTM3  KLK3                6             2     4\n",
       "GDF15  ERBB2               4             0     4\n",
       "       TGFBR2              3             0     3\n",
       "CALR   P4HB                5             2     3\n",
       "       TAP2                4             1     3\n",
       "       LRP1                3             0     3\n",
       "GDF15  EGFR                3             0     3\n",
       "       PTK2                3             0     3\n",
       "MDK    NCL                 0             2    -2\n",
       "HLA-B  KLRD1               0             2    -2\n",
       "LGALS3 ANXA2               0             2    -2\n",
       "MDK    SDC1                0             2    -2\n",
       "HLA-B  CD8B                0             2    -2\n",
       "HSPA8  MAPK9               0             2    -2\n",
       "VEGFA  ERBB2               0             2    -2\n",
       "       NFE2L2              0             2    -2\n",
       "HLA-B  CD3G                0             2    -2"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['diff']>=3) | (df['diff']<=-2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n",
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n"
     ]
    }
   ],
   "source": [
    "treatment_info = get_treatment_info()\n",
    "\n",
    "dat_concat.obs['treatment_status'] = dat_concat.obs['sample_id'].map(treatment_info).astype('category')\n",
    "dat_concat = dat_concat[dat_concat.obs['proximity_analysis'] != 'background']\n",
    "dat_concat = dat_concat[(dat_concat.obs['treatment_status'].isin(['untreated','bicalutamide']))]\n",
    "\n",
    "# For tumor\n",
    "tumor_subset = dat_concat[dat_concat.obs['proximity_analysis'] == 'tumor']\n",
    "sc.tl.rank_genes_groups(tumor_subset,groupby='treatment_status')\n",
    "\n",
    "tumor_subset_res_df = sc.get.rank_genes_groups_df(tumor_subset,group='untreated')\n",
    "tumor_subset_res_df = tumor_subset_res_df[(tumor_subset_res_df['logfoldchanges']>=1) &(tumor_subset_res_df['pvals_adj']<0.05)].sort_values('pvals_adj',ascending=True).reset_index(drop=True)\n",
    "\n",
    "# Remove the lfc>1 condition\n",
    "# \n",
    "# For stroma\n",
    "target_subset = dat_concat[dat_concat.obs['proximity_analysis'] == 'proximal_'+target]\n",
    "sc.tl.rank_genes_groups(target_subset,groupby='treatment_status')\n",
    "\n",
    "target_subset_res_df = sc.get.rank_genes_groups_df(target_subset,group='untreated')\n",
    "target_subset_res_df = target_subset_res_df[(target_subset_res_df['logfoldchanges']>=1) & (target_subset_res_df['pvals_adj']<0.05)].sort_values('pvals_adj',ascending=True).reset_index(drop=True)\n",
    "# Remove the lfc>1 condition\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GDF15</td>\n",
       "      <td>26.265862</td>\n",
       "      <td>1.999140</td>\n",
       "      <td>5.177032e-126</td>\n",
       "      <td>7.744323e-124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ALDH1A3</td>\n",
       "      <td>21.899084</td>\n",
       "      <td>1.276422</td>\n",
       "      <td>3.913762e-91</td>\n",
       "      <td>3.682136e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SPON2</td>\n",
       "      <td>20.418461</td>\n",
       "      <td>1.678994</td>\n",
       "      <td>1.163825e-83</td>\n",
       "      <td>9.891851e-82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TFF3</td>\n",
       "      <td>19.925171</td>\n",
       "      <td>1.576049</td>\n",
       "      <td>1.137101e-79</td>\n",
       "      <td>8.952579e-78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>LRIG1</td>\n",
       "      <td>15.615120</td>\n",
       "      <td>1.221264</td>\n",
       "      <td>5.375101e-51</td>\n",
       "      <td>2.466446e-49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>ADM</td>\n",
       "      <td>6.099532</td>\n",
       "      <td>1.187929</td>\n",
       "      <td>1.322846e-09</td>\n",
       "      <td>8.790964e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>EGF</td>\n",
       "      <td>3.458113</td>\n",
       "      <td>1.210616</td>\n",
       "      <td>5.569990e-04</td>\n",
       "      <td>1.526873e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       names     scores  logfoldchanges          pvals      pvals_adj\n",
       "6      GDF15  26.265862        1.999140  5.177032e-126  7.744323e-124\n",
       "12   ALDH1A3  21.899084        1.276422   3.913762e-91   3.682136e-89\n",
       "16     SPON2  20.418461        1.678994   1.163825e-83   9.891851e-82\n",
       "17      TFF3  19.925171        1.576049   1.137101e-79   8.952579e-78\n",
       "38     LRIG1  15.615120        1.221264   5.375101e-51   2.466446e-49\n",
       "150      ADM   6.099532        1.187929   1.322846e-09   8.790964e-09\n",
       "243      EGF   3.458113        1.210616   5.569990e-04   1.526873e-03"
      ]
     },
     "execution_count": 292,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_de_res = tumor_subset_res_df[tumor_subset_res_df['names'].isin(ligands)][:10]\n",
    "tumor_de_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>ADRB1</td>\n",
       "      <td>9.275340</td>\n",
       "      <td>1.098906</td>\n",
       "      <td>5.406398e-20</td>\n",
       "      <td>5.847744e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>RET</td>\n",
       "      <td>6.301174</td>\n",
       "      <td>1.395112</td>\n",
       "      <td>3.645534e-10</td>\n",
       "      <td>1.680023e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    names    scores  logfoldchanges         pvals     pvals_adj\n",
       "46  ADRB1  9.275340        1.098906  5.406398e-20  5.847744e-19\n",
       "75    RET  6.301174        1.395112  3.645534e-10  1.680023e-09"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_subset_res_df[target_subset_res_df['names'].isin(receptors)][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n",
      "WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'\n"
     ]
    }
   ],
   "source": [
    "# For untreated\n",
    "untreated_subset = dat_concat[dat_concat.obs['treatment_status'] == 'untreated']\n",
    "sc.tl.rank_genes_groups(untreated_subset,groupby='proximity_analysis')\n",
    "\n",
    "untreated_subset_res_df = sc.get.rank_genes_groups_df(untreated_subset,group='tumor')\n",
    "untreated_subset_res_df = untreated_subset_res_df[(target_subset_res_df['logfoldchanges']>=1) &(untreated_subset_res_df['pvals_adj']<0.05)].sort_values('pvals_adj',ascending=True).reset_index(drop=True)\n",
    "\n",
    "# For treated\n",
    "treated_subset = dat_concat[dat_concat.obs['treatment_status'] == 'bicalutamide']\n",
    "sc.tl.rank_genes_groups(treated_subset,groupby='proximity_analysis')\n",
    "\n",
    "treated_subset_res_df = sc.get.rank_genes_groups_df(treated_subset,group='tumor')\n",
    "treated_subset_res_df = treated_subset_res_df[(target_subset_res_df['logfoldchanges']>=1) &(treated_subset_res_df['pvals_adj']<0.05)].sort_values('pvals_adj',ascending=True).reset_index(drop=True)\n",
    "# Remove the lfc>1 condition\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TPT1</td>\n",
       "      <td>20.795767</td>\n",
       "      <td>0.488898</td>\n",
       "      <td>4.444382e-89</td>\n",
       "      <td>5.540292e-86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>ODC1</td>\n",
       "      <td>17.175632</td>\n",
       "      <td>0.967804</td>\n",
       "      <td>1.011424e-62</td>\n",
       "      <td>2.801832e-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>PDLIM3</td>\n",
       "      <td>16.726690</td>\n",
       "      <td>1.079383</td>\n",
       "      <td>1.119342e-59</td>\n",
       "      <td>2.576036e-57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>TRGC1</td>\n",
       "      <td>16.466509</td>\n",
       "      <td>0.883722</td>\n",
       "      <td>2.958039e-58</td>\n",
       "      <td>6.604375e-56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>AGR2</td>\n",
       "      <td>15.940055</td>\n",
       "      <td>1.009178</td>\n",
       "      <td>7.006121e-55</td>\n",
       "      <td>1.416278e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>TMEM178A</td>\n",
       "      <td>16.045340</td>\n",
       "      <td>1.618261</td>\n",
       "      <td>1.197280e-54</td>\n",
       "      <td>2.388015e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>SERHL2</td>\n",
       "      <td>15.035718</td>\n",
       "      <td>1.153568</td>\n",
       "      <td>9.153812e-49</td>\n",
       "      <td>1.538560e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>15.138084</td>\n",
       "      <td>2.333115</td>\n",
       "      <td>1.298183e-48</td>\n",
       "      <td>2.157725e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>MYO6</td>\n",
       "      <td>14.582057</td>\n",
       "      <td>0.948284</td>\n",
       "      <td>2.786766e-46</td>\n",
       "      <td>4.434812e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>AP001816.1</td>\n",
       "      <td>13.855103</td>\n",
       "      <td>0.911142</td>\n",
       "      <td>4.376042e-42</td>\n",
       "      <td>6.005616e-40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         names     scores  logfoldchanges         pvals     pvals_adj\n",
       "8         TPT1  20.795767        0.488898  4.444382e-89  5.540292e-86\n",
       "42        ODC1  17.175632        0.967804  1.011424e-62  2.801832e-60\n",
       "52      PDLIM3  16.726690        1.079383  1.119342e-59  2.576036e-57\n",
       "53       TRGC1  16.466509        0.883722  2.958039e-58  6.604375e-56\n",
       "56        AGR2  15.940055        1.009178  7.006121e-55  1.416278e-52\n",
       "57    TMEM178A  16.045340        1.618261  1.197280e-54  2.388015e-52\n",
       "68      SERHL2  15.035718        1.153568  9.153812e-49  1.538560e-46\n",
       "69        FLI1  15.138084        2.333115  1.298183e-48  2.157725e-46\n",
       "72        MYO6  14.582057        0.948284  2.786766e-46  4.434812e-44\n",
       "83  AP001816.1  13.855103        0.911142  4.376042e-42  6.005616e-40"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_of_interest_df = untreated_subset_res_df[~(untreated_subset_res_df['names'].isin(treated_subset_res_df['names']))][:10]\n",
    "genes_of_interest_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_of_interest = genes_of_interest_df['names'].reset_index(drop=True)\n",
    "df = pd.DataFrame(dat_concat[:,genes_of_interest].layers['counts'].toarray(),columns=genes_of_interest,index = dat_concat.obs_names)\n",
    "df = pd.merge(df,dat_concat.obs[['proximity_analysis','treatment_status']],left_index=True,right_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARcAAACnCAYAAAA7bhgVAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAAsTAAALEwEAmpwYAAAhUUlEQVR4nO3de1xUZRoH8N8Mw1WIm4qKWeoK3kjQ0UoDBREwuWkYyEdzIy9EWn28FF6WvJVhrG7rJbVQP2Vk0UWDNLSgWtm2xS1JRJdCQFFC4yZXGZh3/yDOcrgMAzNnzhx4vv84z5yZ9zxzZJ45l/e8r4wxxkAIIXomFzsBQkjfRMWFECIIKi6EEEFQcSGECIKKCyFEEFRcCCGCMHhx2bdvH1xdXZGXlwcAuHjxIoKDg+Hv74+oqCiUlZUZOiVCiAAMWlwuX76MixcvwtnZGQCgVquxfv16xMXFIS0tDUqlEgkJCYZMiRAiEIWhVtTY2Iht27bhr3/9K5566ikAQE5ODszNzaFUKgEAERERmD17Nnbu3KlVm2q1GrW1tTA1NYVMJhMsd0JIR4wxqFQqDBgwAHJ5x/0UgxWXN998E8HBwRg+fDj3XElJCYYNG8bFDg4OUKvVqKyshJ2dXbdt1tbWcodXhBBxuLi4wMbGpsPzBikuP/30E3JycrBu3Tq9tmtqagqg5cOZmZnptW1tVFZWYt26dWhqaoJCoUBCQoJWRVFM+/fvR1ZWFhdPnToVzz33nIgZiWv58uVQqVRcbGpqirffflvEjLoXFRUFtVrNxXK5HEeOHDF4Ho2NjcjLy+O+h+0ZpLhkZWUhPz8fs2fPBgD89ttveOaZZ7BkyRLcunWLe115eTnkcrnWX9DWQyEzMzOYm5vrPe/uJCcno7KykisuycnJWLVqlcHz6Inc3FxUVlbyYjG2nbG4c+dOh+eMfXtUVlZ2KC5i5tzVKQmDnNBdsWIFzp8/j/T0dKSnp2PIkCFITEzEsmXL0NDQgAsXLgAATpw4gYCAAEOkpBcZGRloamoCADQ1NSEjI0PkjLp38+ZNjTExfm0LS2exsRC1n4tcLseuXbuwdetW+Pn5ISsrC2vXrhUzpR559NFHefH06dNFykR7rVfquoqJ8Wu/p2CsFzMMdkK3rfT0dO7x5MmTkZKSIkYaeieF0StGjhzJ21sZNWqUiNmIz87OjneYaG9vL14yWmr/d2asf3fUQ1cH33//vcbYGP3444+8+D//+Y9ImRiHtoUFACoqKsRJpA+i4qIDb29vmJiYAABMTEzg7e0tckbdGzlyJC8ePXq0SJkYh/YXD6Sw5yIVVFx0EBkZySsukZGRImfUvcuXL/PiS5cuiZSJcaA9F+FQcdGBg4MD5syZA5lMhjlz5sDBwUHslAgxGlRcdDR37lxYWlri8ccfFzsVrUjlSgORPiouOjpz5gzq6+tx+vRpsVPRSkxMDC829k5/QqNiKxwqLjooLy/HuXPnwBjDuXPnUF5eLnZK3WrfF+eRRx4RKRPjIJXLulIkSj+XviIpKYnrHalWq5GUlGT0ewJJSUmQyWRgjEEmk0kiZyG1bou2sVi++uornD17tlfvfemllzQu9/Pzg6+vb6/a7i3ac9GBFLv/Z2RkcF8mxpgkchZS+5vuuroJz5g4OjpqjI0F7bnowNvbG2lpadyNi1Lo5/Loo4/i66+/5mIp3LIgpDFjxvAuz48ZM0a0XHx9fbXeu5g7dy73+Pjx40KlpBPac9FBZGQkN0iOXC6XRD+X9vr7OYbc3FyNsbFq3VuJiIgQOZOuUXHRgRT7uUjxlgUhSfWE7rBhw+Dm5oalS5eKnUqXqLjoKDIyEhMmTJDMXsvEiRM1xoToCxWXfqZ99//2MSH6QsVFR0lJSbh8+TKSkpLETkUrtbW1GmNC9IWKiw6k2InO0tKSF1tZWYmUCenrqLjooLNOdMbO2tpaY0yIvlBx0YEUO9G1H5D69u3bImVC+joqLjrw9vaGQtHSD1EqnegGDRrEiwcPHixSJqSvM1hxiYmJQXBwMEJDQxEZGYkrV64AAAoKChAeHg5/f3+Eh4ejsLDQUCnpTIqd6GpqajTG/Q3dFS0cgxWX+Ph4fP755zh58iSioqKwceNGAMArr7yCyMhIpKWlITIyEnFxcYZKSWdS7ERXX1/Pi+vq6kTKxDhItROdFBisuLSd7rGmpgYymQxlZWXIzc1FYGAgACAwMBC5ubmSuOrSavr06ZDJZJgxY4bYqWiFrhYRQzHojYubNm1CZmYmGGN45513UFJSAicnJ944tIMHD0ZJSYkk9gIA4NChQ1Cr1Th06BAOHjwodjrdsra25u290NUiIhSDFpdXX30VAHDy5Ens2rULL7zwgl7azcnJ0Us7PVVSUoLr168DAIqKipCamoqhQ4eKkou2Orta1N+nF2lPCtujuroagHHnKsqQC6GhoYiLi8OQIUNQWlqK5uZmmJiYoLm5Gbdv3+7xF3TixImizJW7cuVKXpyamiqJvZf2pkyZInYKRkUK2+PDDz8EIG6u9+7d0/jDbpBzLrW1tSgpKeHi9PR02NrawtHREePGjUNqaiqAli/nuHHjJHNI1LrX0qqoqEikTAgxPgbZc6mvr8cLL7yA+vp6yOVy2Nra4uDBg5DJZNiyZQtiY2Nx4MAB3HfffYiPjzdESnoxePBgXic06jMiPaamplCpVLyY6IdBisvAgQPx0Ucfdbps9OjRSE5ONkQaetd63NtVbIxsbW1RVVXFxe1nHOxv2haWzmLSe9RDVwft+4y0j41R28ICdJxxkBB9oeKig/Z9RtrHxkgqgzsT6aPiogMp3mHcfi5kmhuZCIWKiw7a9xlpHxuj1iEiuooJ0RcqLjoYMWIEL37ggQdEykR7rb2hu4oJ0RcqLjpoP8tdd7PeGYP2V4f6+9UiIhwqLjoYPXo0bzyXUaNGiZxR98rKyjTGhOgLFRcd5Ofn80aiu3btmsgZdU+Kh3JEmqi46GDXrl0aY2MkxUM5Ik1UXHQgxXuL7O3tuccymYzOuRDBaCwudXV12LJlC4KDg7F27doOX6agoCBBkzN2UhwiMSkpiRuaUyaTSWLGAiJNGovLzp07cfPmTaxevRoODg5YuHAhfvjhB255cXGx4AkaMykOkZiRkcGbDkUKMxYQadJ442JGRgbOnDkDGxsbzJkzB7NmzcKLL76I119/HTNnzjRUjkbL0tKSdz+RFLr/u7i44OLFi1w8duxY8ZIhfZrGPZd79+7BzMyMi2fMmIG33noLGzduxJkzZyRxGCCktuMCdxYbo0uXLvHi7OxskTIhfZ3G4jJmzBhkZWXxnnN3d8fbb7+NHTt2oKGhQdDkjF37CcWkMMFYc3OzxpgQfdFYXFauXNnhFn0AGD9+PI4ePYqQkBDBEpOCIUOGaIyNUWunv65iQvRFY3GZOXMm5s2b1+kyFxcXbNu2TZCkpEoKh4lr167lxevXrxcpE9LX9aqfS2NjI9599134+vrqOx9J+e2333hx23GCjdWsWbO4IiiTyeDl5SVyRqSv0lhcrl27hkWLFsHDwwPz589HXl4e0tLSMHv2bHz++ed4+eWXDZWnUXJ2dtYYG6Py8nLePFFSmoCOSIvGA+5XX30VDz74IKKjo5GamoqYmBhYWFggPj4e06dPN1SORmvkyJG4efMmF0vhxsWkpCSuPw5jDElJSVi1apXIWZG+SGNxycnJwT/+8Q+YmZlh6tSpmDJlCjIyMnp84rKiogIvvfQSrl+/DjMzMzzwwAPYtm0bHBwccPHiRcTFxeHevXtwdnbGG2+8IZmhF3/88UdebMwTVLXKyMjgrhA1NzcjIyODigsRhMbDIpVKxfVzsbKygo2NTa+uiMhkMixbtgxpaWlISUnB/fffj4SEBKjVaqxfvx5xcXFIS0uDUqlEQkJC7z6JCFxcXHixq6urSJlob+LEibzYzc1NpExIX6dxz6WxsRFvvvkmFzc0NPBiAFpNyWpnZ4eHH36Yi93d3fHBBx8gJycH5ubmUCqVAICIiAjMnj0bO3fu7NGHEEv72ebad1AzRlLMmUiTxuISFBTEuyIyb968DldIekqtVuODDz6Aj48PSkpKMGzYMG6Zg4MD1Go1Kisre3S3rlhzRbeO5dI2NvZDo7q6ug6xsedsaFLYHpKfK1qIPYjt27fDysoKixcvxrlz5/TSplhzRcvlct4A13K53OjnGW5/P5SVlZXR52xoUtgeUpgrutvumSqVipvi8sKFC7w7fz08PHrUwzM+Ph5FRUU4ePAg5HI5hg4dilu3bnHLy8vLIZfLJTPGiKmpKe7du8eLjZ21tTWvuEhhOhQiTRorQ1JSEn766Se88cYbAIBnnnmG++I3NDRg3bp1WLhwoVYr2r17N3JycnD48GHuJPHEiRPR0NCACxcuQKlU4sSJEwgICNDh4xhW28LSWWyM2k9/IoX7oYg0aSwup06dwtatW7nYzMwM3377LQDgypUr2LJli1bF5ZdffsGhQ4fw4IMPIiIiAgAwfPhw7N+/H7t27cIrr7zCuxRNhGNtbY2amhpe3Bd99dVXOHv2bK/e293Qn35+fv2+d7o2NBaX4uJi3ngfo0eP5h6PHTsWN27c0GolY8aMwX//+99Ol02ePBkpKSlatUN019lJ6P7M1taWd3Oura2tiNn0LRqLS11dHerq6mBlZQUAOHHiBG+ZFCZeF5IU9wIeeugh/Pvf/+biSZMmiZiNcHx9fbXeu5g7dy73uO3fONFNt+O5ZGZmdrrs/Pnz+NOf/iRIUlKxYcMGXrxx40aRMtHe5cuXebFYl/GNSeveir+/v8iZ9C0a91yWLl2KrVu3QiaTwcfHh7v0+vXXX2P79u2IjY01VJ5GafLkybzYw8NDpEy0V1tbqzHuj1rncnrxxRfFTaSP0Vhc5s2bh9LSUqxfvx4qlQp2dnaorKyEqakpVq1ahccff9xQeRql1r4GrZKTk7W+eiYWc3Nz3lUtMfoHkf6h204qUVFRePLJJ/Hjjz9yPWctLS2Rnp4OLy8vnD9/3hB5GqVjx47x4iNHjhh9caETuoZ18OBBQWbibG1TiEntRo0ahejoaJ3b0aoHnLW1NSZOnIiUlBQcPXoUV69ehVKpxKZNm3ROgBgWjaFrWNeuXUNedjacVCq9tmv+x9xTVRcu6LXdUj12BNVYXFQqFdLT0/HZZ5/h/PnzGDFiBObNm4dbt27hb3/7m2SGRiD/Z2JiwisorQNHEeE4qVRYLJFBuY47OOitLY3FZcaMGZDJZFiwYAFWr16NCRMmAAA++OADvSVADEuhUPCKCw3QTYSi8VK0q6srqqurkZ2djUuXLnU6EwCRFineskCkSWNxee+993Du3DnMmDEDR44cwYwZMxAdHY26ujo6EQhpzhVtYWGhMSZEX7od/d/Z2RnPPfcczp49i2PHjmHQoEGQy+UIDg7Grl27DJGj0ZJicSHEUHp0wK1UKqFUKrF582acO3cOJ0+eFCgtaWg7lktnsTFqP0tmf581kwinV2fzzM3NERgYiMDAQH3nQwRmZmaGxsZGXkyIEHo1KRqRrraFpbOYEH2h4kIIEQQVF0KIIKi49DPtx/mVwri/RJqouPQzUpzIjUgTFZd+hgaLIoZikOISHx8PHx8fuLq6Ii8vj3u+oKAA4eHh8Pf3R3h4OAoLCw2RDiHEAAxSXGbPno33338fzs7OvOdfeeUVREZGIi0tDZGRkYiLizNEOv0a9SomhmKQW2Jb54Juq6ysDLm5uTh69CgAIDAwENu3b0d5eTkc9Hjbd29oOy3F4MGDefP+DB482OinpYiJicH+/fu5eNWqVaLlQvo20e63LykpgZOTEzeeiImJCQYPHoySkpIeFxd9nzcoLCzk5uLVpHVWhLZxd+8rLCwUdX7foUOH8mInJyejnm/YEIScd1mbvyNjU11drZdt0ScG89D3XNFTpkzBsmXLtHrt0qVLcfv2bURFRRn9EJcAkJ+fz4vt7e0xatQokbIxDkLOu/zhhx9CagOV2NjYaLUtupsrWrSrRUOHDkVpaSk3cFFzczNu377d4ZfV2Dk5OcHNzU0ShQVAhzvZ+/ud7UQ4ohUXR0dHjBs3DqmpqQCA1NRUjBs3TvTzLX3d9evXeXFRUZFImZC+ziDFZceOHfDy8sJvv/2Gp59+GvPmzQMAbNmyBcePH4e/vz+OHz/Om5eaCKP9uMc0DjIRikHOuWzevBmbN2/u8Pzo0aORnJxsiBTIHyorKzXGxkyK03QUFBRAZWqq14GvhVRqagp1RYVe2uoTJ3SJ9qQ8tci1a9eQ+99cWDpa6rXdJkXLkK0Fvxfotd36snqYwAT9tScRFZd+RqFQ8MY/ltro/5aOlhgb6NL9C43A1dQ8sLsMjtXVkppaxNbeXi9tSesvi3RJ245/Q4YMQXFxMS829o5/RJroxsV+xr7dr1L7mBB9oT2XPsLX11frvYvly5ejuLgYGzZsgJeXl8CZkf6Kiks/ZG9vD3t7eyosRFB0WEQIEQQVF0KIIKi4EEIEQcWFECKIfnNCV4pdx0eNGoXo6Gi9t0uIIfSb4iLFruOESFm/KS6A9LqOEyJl/aq4SM2tW7cEOdyiQznDKhXgrugaecvpUmu1Wq/tlpqawlZPbVFxMWINDQ3Iy86Gk0ql13bN//jDrLpwQa/tltLsjR1YWFhgmADDiJb98QPhrOe2bQG9DXtKxcXIOalUkrqjlvANGzZMkKFEW/c6jXmYUiouRDIqKipQV1YnmfNRdWV1qDDRz8BLUkT9XAghgug3ey5S/NWTNcsEORkoFH0OkdgZe3t7lNwu0Xu7qrqWc1qmVvo9ZySDrF8PaWEUxaWgoACxsbGorKyEnZ0d4uPj8eCDD4qdFjEyQs2v1Hr1bOTAkfpteKBwOUuBURSX1jmjQ0JCcOrUKcTFxeHdd9/V6zrs7e1R2VwpqX4u7C6DY12dpE7o6muIxM4IdYlbCidHpUj04qLLnNGMMQBAY2Njt+uxsrKCeZM5Sr4t7fa1qnoVmhqaun1dbygsFDC17H7327zJHOa25mgwMcFJLb6wdXI56uTCnEKzUqthpUV/igaFAkOtrHDv3j1B8uiJzMxMfPfdd1q9tqysDEDLVDfa8PLywowZM3qbWpeklnPr9671e9ie6MVFlzmjVX/0/8jL6/48SlBQkO7JEq3oe+7u3rC1tRX0/1yIzyjFnIGW76GFhUWH50UvLroYMGAAXFxcYGpqCpmsv07gQIg4GGNQqVQYMGBAp8tFLy5t54w2MTHp0ZzRcrkcNjY2BsiSENKZzvZYWonez4XmjCakb5Kxrs7GGFB+fj5iY2Nx9+5d3HfffYiPj+/Xl/AI6QuMorgQQvoe0Q+LCCF9ExUXQoggqLgQQgRBxYUQIggqLoQQQVBx6YSrqytqa2s7XRYSEoKGhgZB2m5r7969Wt0z1ROlpaVYsmRJl8u1za2/unv3Lt5++21B2v70009RUNC7GST27t2L+Ph4PWekOyouPXTq1CmNvRL1Zd++fdy9U/ri5OSE9957T69t9id3797FO++80+Xypqbe3+z62WefobCwsNfvN0ZUXLqQmJiIkJAQ+Pv7Iy0tjXu+7a97fn4+oqKiEBQUhKCgIHz22WcAgCNHjuCJJ55AaGgowsPDceXKlU7X0X5PoTXeunUrACAiIgIhISG4e/cuUlJSsHDhQoSGhiI0NBTff/899z4fHx/s2bMH4eHhmDVrFlJSUnDs2DGEhYVhzpw5yMrKAgAUFxfj4Ycf5t539uxZBAQEICQkBPv37+fllp2djSVLlmDBggVYsGABvvnmGx22pvFpvy1a49Z/9+zZg9DQUPj7++PCHwOZb9u2DdXV1QgJCUFERAQAYMmSJXj11Vfx5JNP4tlnnwUAHD58GGFhYZg/fz6io6Nx584dAMD333+P8PBwhIaGIigoCF988QUA4JNPPkFOTg527NiBkJAQ/POf/9TYTnV1NZ5//nkEBARgyZIluH79umE2Wk8x0oGLiwvbu3cvY4yx/Px8Nm3aNPb7779zy2pqaphKpWJ+fn7s9OnT3PvKy8sZY4yVlZVxz2VmZrKFCxfy2q6pqenwuLtl5eXlTK1Wczl5enpyy7y9vdnrr7/OGGMsOzubTZo0iR0/fpwxxtgXX3zBIiIiGGOM3bhxg02bNo0xxtidO3fYtGnTWH5+PmOMscOHD3PrrKqqYiEhIay0tJQxxlhpaSnz9PRkVVVVvdmcRqnttmgb37hxg7m4uLD09HTGGGOnTp1i4eHhnb6HMcYWL17MVq5cyVQqFWOMsZMnT7LNmzez5uZmxhhj77//PluzZg1jjLHKykrW1NTEGGvZ/p6enqyyspJrp3Wd3bWzc+dOFhsbyxhr+VubOXMm9/9vTES/cdFYLVy4EEDLSGLjx4/HxYsXMXv2bG55QUEBmpqaMHfuXO651iENc3JycOjQIVRVVUEmk+lld/fGjRtYu3YtSktLoVAo8Pvvv+POnTsYNGgQAODxxx8HAEyYMAH19fVcXhMnTuz0ly07Oxvjx4/nbrMIDw9HQkICAOCnn35CcXExli9fzr1eJpOhqKgIbm5uOn8WY2dlZQVvb28AgLu7e7fnM4KCgqBQtHyV0tPTkZOTg/nz5wMAmpubYW1tDQAoLy/Hxo0bUVRUBBMTE1RVVaGgoADu7u4d2tTUzg8//IDNmzcDABwcHDBnzhzdP7QAqLjoWWNjI1544QUcP34cEyZMQGlpKby8vDp9rYmJCTfQTncDLK1ZswaxsbHw9fWFWq3GpEmTeO8xNzfn2mwby+XyHp8LYIzB1dUV77//fo/eJyUKhYI3yFHbbWlmZsY91mb7WVlZcY8ZY3j22WcRFhbW4XVbtmyBj48P9u3bB5lMBn9//y7/3zW1IxV0zqULn3zyCQCgsLAQubm5HX5dRo4cCYVCgTNnznDPVVRUoLGxEU1NTdyQEUlJSV2uY8SIEbh06RIAICUlhbdswIABqKmp4eLq6moMHz6cy03XK0nu7u7Izc3l9qqSk5O5ZR4eHigqKsK//vUv7rmff/65yxHHpGjgwIFQqVQoKioCAO6ufE2sra3R0NCgsdj4+PggKSkJVVVVAFp+bK5evQqg5f/Q2dkZMpkMmZmZ3LqBlv/v6upqrdp55JFH8OmnnwJo+Zv76quvevLRDYb2XLrQ3NyM0NBQ1NfXY9u2bXB0dOQtVygUOHDgALZt24YDBw5AJpMhKioKoaGheP755xEWFgY7Ozv4+/t3uY4NGzYgLi4ONjY2CAgI4C2LiorCU089BQsLC7z33nvYsGEDYmJiYGtrC09PT9jZ2en0+RwdHbF9+3ZER0fDwsICfn5+3DJbW1scOHAAb7zxBl577TWoVCrcf//9OHjwYJ8ZlEuhUGDTpk14+umn4eDggFmzZnX7Hjs7O+7kva2tLU6cONHhNaGhoaisrMTixYsBtOyBLFq0CGPHjsXatWuxdetW7N27F25ubnB1deXeFx4ejtdffx2JiYl4+eWXNbYTExODjRs3IiAgAIMGDYJSqdTPRtEzuiuaECIIOiwihAiCigshRBBUXAghgqDiQggRBBUXQoggqLgQQgRBxYUQIggqLhLk4+PD3Tmrb3v37sW6desEaVsbrq6uvJ6rvdHT7RMbG4s9e/botE7SERWXPkaXMUUI0SuxbscmvbNu3Trm6urK3NzcmLu7OzdUwkcffcRmzpzJIiMjGWOMJScns4CAAKZUKllUVBQrLi7m2ti+fTvz8vJiHh4ebP78+SwrK4sxxti3337LJkyYwMaPH8/c3d1ZUFAQY6xlOIDdu3ez8PBw5u7uzlauXMnKy8vZmjVrmIeHB1uwYAG7ceMG1/6vv/7K/vznP7OpU6cyPz8/9sUXX3DLXn75ZbZlyxa2fPly5u7uzsLCwlhRURFjjLHIyEjm4uLCJk2axNzd3Xnva6+srIytWLGCTZkyhU2dOpUtWrSINTc3d7p9GGNs9erVbPr06Wzy5MksMjKS5eXlMcYYO3HiBBs/fjybMGEC99kYaxnyorCwkJf37t27Na6b8FFxkSBvb2+WmZnJGGPc+CPr169ntbW1rL6+np07d475+vqyX3/9lalUKrZ//35uTBLGWsYKKS8vZyqViiUmJrLp06ezhoYGxhhjf//739natWt561u8eDHz9fVlRUVF7O7du2zu3LnMz8+PZWZmMpVKxdavX8+NL1JbW8u8vLzYxx9/zFQqFbt8+TKbNm0a++WXXxhjLV/SadOmsezsbKZSqdiaNWvYiy++yK2r/Ze6KwkJCewvf/kLa2xsZI2NjSwrK4sb76bt9mmVnJzMqqur2b1799iOHTtYcHAwt6xt4egqj7av0bRu8n90WNRHrF69GlZWVrCwsMCJEyewYsUKjB49GgqFAtHR0bhy5Qpu3rwJoGUcYHt7eygUCkRFRaGxsbHb8VsXLFiAESNGwMbGBl5eXrj//vsxffp0KBQKBAQEIDc3FwDwzTffwNnZGU888QQUCgXGjx8Pf39/fPnll1xbvr6+eOihh6BQKBAcHNzlSH2aKBQK3LlzB7du3YKpqSmUSqXGmyrDwsJgbW0NMzMzrF69GlevXuXdhSzkuvsruiu6jxgyZAj3+NatW3jttdd4gxwxxlBaWgpnZ2ckJibi448/xu3btyGTyVBTU4OKigqN7Q8cOJB7bG5uzostLCxQV1cHALh58yZ+/vln3p26zc3NCA4O7rSttu/tiWeeeQb79u1DVFQUgJa7ilesWNHpa5ubm7Fnzx58+eWXKC8vh1ze8ptaUVEBGxsbQdfdn1Fx6SPa/nIOHToU0dHRvC90qwsXLuCdd97BsWPHMGbMGMjlckydOpUbq0XXX+ChQ4di6tSpOHr0qE7tdMfa2hqxsbGIjY1FXl4eli5dCjc3Nzz66KMdXpuSkoKvv/4aR48exfDhw1FdXd3tZ7a0tER9fT0X37lzB05OTj1ed39Gh0USNHDgQNy4caPL5RERETh8+DB++eUXAC2DFLUOalVbWwsTExM4ODigqakJ+/bt4w1K5ejoiJs3b0KtVvcqt1mzZqGwsBAnT56ESqWCSqXCzz//jPz8fL18tlYZGRkoKioCYww2NjYwMTHhikT7Nmpra2FmZgZ7e3vU19dj9+7dvLYcHR1RXFzMe27s2LFITU1Fc3MzvvvuO26Q8+7WTf6PiosErVixAm+99RaUSiVvZoJWc+bMwbJly7BmzRpMnjwZgYGB+O677wAAjz32GDw9PeHv7w8fHx+Ym5tzo+YB4Aatevjhh7nxW3vC2toaiYmJOH36NDw9PfHYY48hISFB65HzVq1ahdjYWCiVSpw+fbrL1xUVFeHpp5+Gh4cHwsPDsWjRIjzyyCMA+NsnMTERoaGhGDZsGDw9PTFv3rwOowqGhYXh119/hVKpRExMDABg06ZNyMjIgFKpREpKCnx9fbVaN/k/GiyKECII2nMhhAiCTugSo3Xw4EEcOnSow/NTpkzROPMhMQ50WEQIEQQdFhFCBEHFhRAiCCouhBBBUHEhhAjifxZKEHJiCwgvAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 288x144 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "g = genes_of_interest[4]\n",
    "# Set the figure size and theme\n",
    "sns.set(rc={'figure.figsize':(4,2)})\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "\n",
    "# Define custom color palette and color codes\n",
    "custom_palette = ['#db1616','#74b572','#f7f774']\n",
    "color_codes_wanted = ['tumor', 'normal', 'stroma']\n",
    "\n",
    "# Define a lambda function to retrieve color from custom palette based on color codes\n",
    "c = lambda x: custom_palette[color_codes_wanted.index(x)]\n",
    "\n",
    "# Create the boxplot using seaborn\n",
    "sns.boxplot(data=df, x=\"treatment_status\", y=g, hue=\"proximity_analysis\", palette = sns.color_palette([c('normal'),c('tumor')],2))\n",
    "\n",
    "# Set the y-axis limits\n",
    "plt.ylim(0,40)#np.quantile(df[g],0.95)\n",
    "\n",
    "# Remove the legend frame\n",
    "plt.legend([],[], frameon=False)\n",
    "\n",
    "# Set the DPI to 200\n",
    "plt.savefig('./plots/output_'+g+'.pdf')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading and inspecting ligand-receptor binding results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tumor --&gt; stroma (11)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(CD99, CD81)</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PPIA, MAPK3)</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PPIA, CXCR4)</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, LRP1)</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(MIF, CD74)</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, ITGAV)</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TNF, MYLK)</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(KIT, CD81)</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(MDK, ITGB1)</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-C, CD81)</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(LAMB2, ITGB1)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(AZGP1, ITGAV)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(SPON2, ITGB1)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HSPA8, MAPK10)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PLA2G2A, ITGB1)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(VEGFA, ITGB1)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, AR)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, SCARF1)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PLA2G2A, ITGA5)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(LAMB2, ITGA5)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TGFB1, DCN)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CIRBP, MAPK3)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALML3, MYLK)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, TAP2)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(VWF, THBS1)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TP53, THBS1)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-DRB1, CD81)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, BDKRB2)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(THBS1, ITGA5)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-C, CD3G)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tumor --> stroma (11)\n",
       "(CD99, CD81)                          9\n",
       "(PPIA, MAPK3)                         9\n",
       "(PPIA, CXCR4)                         9\n",
       "(CALR, LRP1)                          9\n",
       "(MIF, CD74)                           8\n",
       "(CALR, ITGAV)                         8\n",
       "(TNF, MYLK)                           8\n",
       "(KIT, CD81)                           7\n",
       "(MDK, ITGB1)                          7\n",
       "(HLA-C, CD81)                         7\n",
       "(LAMB2, ITGB1)                        6\n",
       "(AZGP1, ITGAV)                        6\n",
       "(SPON2, ITGB1)                        6\n",
       "(HSPA8, MAPK10)                       6\n",
       "(PLA2G2A, ITGB1)                      6\n",
       "(VEGFA, ITGB1)                        6\n",
       "(CALR, AR)                            6\n",
       "(CALR, SCARF1)                        5\n",
       "(PLA2G2A, ITGA5)                      5\n",
       "(LAMB2, ITGA5)                        5\n",
       "(TGFB1, DCN)                          5\n",
       "(CIRBP, MAPK3)                        5\n",
       "(CALML3, MYLK)                        5\n",
       "(CALR, TAP2)                          5\n",
       "(VWF, THBS1)                          5\n",
       "(TP53, THBS1)                         5\n",
       "(HLA-DRB1, CD81)                      5\n",
       "(CALR, BDKRB2)                        5\n",
       "(THBS1, ITGA5)                        4\n",
       "(HLA-C, CD3G)                         4"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = 'tumor'\n",
    "target = 'stroma'\n",
    "\n",
    "ligand_receptor_dict = load_from_pickle('data/'+source+'_to_'+target+'_ligand_receptor_proximity_interactions.pickle')\n",
    "\n",
    "interface_interaction = most_common_interactions(ligand_receptor_dict)\n",
    "df = pd.DataFrame(interface_interaction,columns=[\n",
    "    source +' --> ' + target + ' ('+str(len(ligand_receptor_dict))+')'])\n",
    "df[:30]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tumor --&gt; normal (8)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(CMTM3, KLK3)</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, P4HB)</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, PDIA3)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(MDK, TSPAN1)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, AR)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PPIA, MAPK3)</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(GDF15, ERBB3)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(PPIA, CXCR4)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TGFB1, P4HB)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(AZGP1, ITGAV)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, TAP2)</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TYK2, RACK1)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, SCARF1)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(CALR, ITGA3)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-A, APLP2)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TP53, CTSD)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(MIF, CD74)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(GDF15, ERBB2)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(WNT11, GNAS)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HSPA8, LDLR)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(NPY, DPP4)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HSPA8, MAPK10)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-C, DDR1)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(WNT4, GNAS)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TP53, HSP90AB1)</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(GDF15, TGFBR2)</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-A, ERBB2)</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(EGF, HSP90AB1)</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(HLA-B, CD8B2)</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(MDK, VEGFA)</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tumor --> normal (8)\n",
       "(CMTM3, KLK3)                        8\n",
       "(CALR, P4HB)                         7\n",
       "(CALR, PDIA3)                        6\n",
       "(MDK, TSPAN1)                        6\n",
       "(CALR, AR)                           6\n",
       "(PPIA, MAPK3)                        6\n",
       "(GDF15, ERBB3)                       5\n",
       "(PPIA, CXCR4)                        5\n",
       "(TGFB1, P4HB)                        5\n",
       "(AZGP1, ITGAV)                       5\n",
       "(CALR, TAP2)                         5\n",
       "(TYK2, RACK1)                        4\n",
       "(CALR, SCARF1)                       4\n",
       "(CALR, ITGA3)                        4\n",
       "(HLA-A, APLP2)                       4\n",
       "(TP53, CTSD)                         4\n",
       "(MIF, CD74)                          4\n",
       "(GDF15, ERBB2)                       4\n",
       "(WNT11, GNAS)                        4\n",
       "(HSPA8, LDLR)                        4\n",
       "(NPY, DPP4)                          4\n",
       "(HSPA8, MAPK10)                      4\n",
       "(HLA-C, DDR1)                        4\n",
       "(WNT4, GNAS)                         4\n",
       "(TP53, HSP90AB1)                     4\n",
       "(GDF15, TGFBR2)                      3\n",
       "(HLA-A, ERBB2)                       3\n",
       "(EGF, HSP90AB1)                      3\n",
       "(HLA-B, CD8B2)                       3\n",
       "(MDK, VEGFA)                         3"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = 'tumor'\n",
    "target = 'normal'\n",
    "\n",
    "ligand_receptor_dict = load_from_pickle('data/'+source+'_to_'+target+'_ligand_receptor_proximity_interactions.pickle')\n",
    "\n",
    "interface_interaction = most_common_interactions(ligand_receptor_dict)\n",
    "df = pd.DataFrame(interface_interaction,columns=[\n",
    "    source +' --> ' + target + ' ('+str(len(ligand_receptor_dict))+')'])\n",
    "df[:30]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "contingency_table = np.array([[5, 2], [3, 9]])\n",
    "print(contingency_table)\n",
    "odds_ratio, p_value = fisher_exact(contingency_table)\n",
    "\n",
    "print(\"Odds Ratio:\", odds_ratio)\n",
    "print(\"p-value:\", p_value)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "squidpy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
